definitions/npm/react-router-dom_v4.x.x/flow_v0.53.x-/react-router-dom_v4.x.x.js

```diff
@@ -73,16 +73,16 @@ declare module "react-router-dom" {
   };
 
   declare export type Match = {
-    params: { [key: string]: ?string },
+    params: { [key: string]: string },
     isExact: boolean,
     path: string,
     url: string
   };
 
   declare export type ContextRouter = {|
-    history: RouterHistory,
-    location: Location,
-    match: Match
+    history: RouterHistory | void,
+    location: Location | void,
+    match: Match | void,
```

Comment:
I think the convention is to use [optional types](https://flow.org/en/docs/types/objects/#toc-optional-object-type-properties) if the property may not exist.

For example,

```js
history: RouterHistory | void,
```

would be

```js
history?: RouterHistory,
```

Reply:
The [`$Diff`](https://flow.org/en/docs/types/utilities/#toc-diff) documentation and [HOC](https://flow.org/en/docs/react/hoc/#toc-injecting-props-with-a-higher-order-component) guide seems to prefer the `| void` notation. I also found lots of definitions using the same pattern so I went with it, happy to change it.

Reply:
Oh, it's perfectly fine with me, I just thought the `?` was more convenient so just wanted to mention it in case you didn't know. Think you just need to get tests to pass.

Reply:
[Both seem to work fine](https://flow.org/try/#0MYewdgzgLgBAhgLhgEgCIEsBmmA8BvAXwBoY8ZMQQkwBXAWwCMBTAJxgB8YA3EdAExgEAfDAC8pAgChQkWAyRosuQiTIUqMAPy1GrQSPGFJQA), and I'd prefer using maybe types, i.e.

```
history: ?RouterHistory
```

@saadq note that there's a difference between optional props (`foo?: number`) and maybe types (`foo: ?number`). We want the latter.

@omeid if you're willing to change it, please do so. I'm fine with leaving it in if you hate it though. 😉

Reply:
Also checking with flow. [Issue](https://github.com/facebook/flow/issues/5902)

Reply:
Thanks @AndrewSouthpaw but I believe we do not want Maybe Types, the props in question here is never null.

And as for `property? : T` vs  `property: T | void`, they're actually handled differently, the first one also includes undefined, so you get weird behavior with diff.

```jsx
 // the error is not caught!
const a1: $Diff<{|p: number|}, {p?: number}> = {p: 1}

 // errors as expected: 
 // Cannot assign object literal to `a2` because property `p` is missing in
 // object type but exists in object literal.
const a2: $Diff<{|p: number|}, {p: number | void}> = {p: 2}
```

Reply:
Yep, agreed.